File: solutions/TypeScript/[0056_Medium] Merge Intervals/merge.ts
function quickSort(arr: number[][]) {
  function swap(items: number [][], leftIndex: number, rightIndex: number) {
    const temp = items[leftIndex];
    items[leftIndex] = items[rightIndex];
    items[rightIndex] = temp;
  }

  function partition(items: number [][], left: number, right: number) {
    const pivot = items[Math.floor((right + left) / 2)][0];
    let i = left;
    let j = right;

    while (i <= j) {
      while (items[i][0] < pivot) {
        i++;
      }
      while (items[j][0] > pivot) {
        j--;
      }
      if (i <= j) {
        swap(items, i, j);
        i++;
        j--;
      }
    }
    return i;
  }

  function sort(items: number [][], left: number, right: number) {
    let index: number;

    if (items.length > 1) {
      index = partition(items, left, right); // index returned from partition
      if (left < index - 1) { // more elements on the left side of the pivot
        sort(items, left, index - 1);
      }
      if (index < right) { // more elements on the right side of the pivot
        sort(items, index, right);
      }
    }

    return items;
  }

  return sort(arr, 0, arr.length - 1);
}

function merge(intervals: number[][]): number[][] {
  quickSort(intervals);

  let min = intervals[0][0];
  let max = intervals[0][1];

  const result: number[][] = [];

  for (let i = 1; i < intervals.length; i++) {
    if (intervals[i][0] <= max) {
      min = Math.min(min, intervals[i][0]);
      max = Math.max(max, intervals[i][1]);
    } else {
      result.push([min, max]);
      min = intervals[i][0];
      max = intervals[i][1];
    }
  }

  result.push([min, max]);

  return result;
}

File: solutions/Java/[0056_Medium] Merge Intervals/merge.java
class Solution {
  int partition(int[][] arr, int low, int high) {
    int pivot = arr[(high + low) / 2][0];
    int i = low;
    int j = high;

    while (i <= j) {
      while (arr[i][0] < pivot) {
        i++;
      }

      while (arr[j][0] > pivot) {
        j--;
      }

      if (i <= j) {
        int[] temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
        i++;
        j--;
      }
    }

    return i;
  }

  void quickSort(int[][] arr, int low, int high) {
    if (arr.length > 1) {
      int middle = partition(arr, low, high);

      if (low < middle - 1) {
        quickSort(arr, low, middle - 1);
      }

      if (middle < high) {
        quickSort(arr, middle, high);
      }
    }
  }

  public int[][] merge(int[][] intervals) {
    quickSort(intervals, 0, intervals.length - 1);

    int min = intervals[0][0];
    int max = intervals[0][1];

    List<int[]> result = new ArrayList<>();

    for (int i = 1; i < intervals.length; i++) {
      if (intervals[i][0] <= max) {
        min = Math.min(min, intervals[i][0]);
        max = Math.max(max, intervals[i][1]);
      } else {
        result.add(new int[]{min, max});
        min = intervals[i][0];
        max = intervals[i][1];
      }
    }

    result.add(new int[]{min, max});

    return result.toArray(new int[result.size()][]);
  }
}
